Connection culling

deweerdt opened the pull request:
This PR implements a mechanism to cleanup idle connections when we reach `max-connections` (as opposed to delaying `accept(2)`).

kazuho commented on lib/http1.c:
```diff
@@ -994,15 +1008,27 @@ static h2o_iovec_t log_request_index(h2o_req_t *req)
 
+static int conn_is_h1(h2o_conn_t *conn)
+{
+    return conn->callbacks->get_sockname == get_sockname;
```
How about promoting `static const h2o_conn_callbacks_t callbacks` defined in `h2o_http1_accept` to a file-level static variable and comparing its address with `conn->callbacks`?

Comparing the address of the callback table is more natural than comparing a particular value of a callback table.

deweerdt commented on lib/http1.c:
```diff
@@ -529,8 +544,7 @@ static void reqread_on_timeout(h2o_timer_t *entry)
 
-    conn->req.http1_is_persistent = 0;
-    close_connection(conn, 1);
+    close_idle_connection(&conn->super);
```
That's a good point. I do assume that `conn->sock->input->size == 0` because `reqread_on_timeout` fired, which means there was nothing to read and we ended up timing out. I'll add the assert.

kazuho replied:
Don't we see a timeout with a non-empty `conn->sock->input` if the client has sent a partial HTTP2 frame?

deweerdt replied:
> Don't we see a timeout with a non-empty conn->sock->input if the client has sent a partial HTTP2 frame?

Would that happen in http1? In a connection upgrade?

kazuho replied:
My bad, please ignore my comment here.

deweerdt commented:
> Would it be preferable to start closing idle connections before we reach max_connections?

Yes, you're correct, it would. I was trying to avoid a soft limit, but i realize now that there's no escaping.

> Would it make sense to guarantee that idle connections would be kept open for a small amount of seconds? 

Yes, it would. Should we hardcode the value to, say, 30s, or would you see a use case to make this configurable?

It sounds like we would end up with one or two additional parameters:
- `soft-connection-limit: <number>`: a number lower than `max-connections` that would have H2O start closing idle connections?
- `soft-connection-limit.min-age: <duration>`: the minimum amount of time that we should wait before considering a connection idle.

deweerdt commented:
@i110 I believe that edf760b addresses the comments you made so far. I will address the suggestions that @kazuho had.

i110 commented on src/main.c:
```diff
@@ -1565,13 +1565,16 @@ static void on_accept(h2o_socket_t *listener, const char *err)
              * _before_ calling `accept`.  In other words t/40max-connections.t may fail.
              */
-            break;
+            if (!h2o_context_close_idle_connections(ctx->accept_ctx.ctx, excess_connections)) {
```
I still see two issues in this line:

1. we should pass `excess_connections + 1` to ensure that there will be at least one slot for the incoming connection
2. for now `h2o_context_close_idle_connections` returns the number of closed connections, so we should compare that with `excess_connections + 1`

Thus I think the following is the correct one:
```c
if (h2o_context_close_idle_connections(ctx->accept_ctx.ctx, excess_connections + 1) < excess_connections + 1) {
```

i110 replied:
And also I realized that it's dangerous to pass `0` to `h2o_context_close_idle_connections` mistakenly like this case, because it will try to close all idle connections if we do so. How about adding `assert(max_connections_to_close > 0)` in the beginning of that function?

deweerdt commented:
@i110 @kazuho I believe that 5f46d7e addresses all your comments so far. In introduces `soft-connection-limit`, which set limits at which H2O will try to close connections older than `soft-connection-limit-min-age`seconds

kazuho commented on src/main.c:
```diff
@@ -1565,14 +1579,21 @@ static void on_accept(h2o_socket_t *listener, const char *err)
         h2o_socket_t *sock;
-        if (num_connections(0) >= conf.max_connections) {
+        if (conf.soft_connection_limit > 0) {
+            int excess_connections = num_connections(0) - conf.soft_connection_limit;
```
Should we better divide the number with `conn.num_threads`? Otherwise I think we'd be retiring connections unfairly aggressively on the thread that hits this condition first.

kazuho commented on src/main.c:
```diff
@@ -1565,14 +1579,21 @@ static void on_accept(h2o_socket_t *listener, const char *err)
     do {
         h2o_socket_t *sock;
-        if (num_connections(0) >= conf.max_connections) {
+        if (conf.soft_connection_limit > 0) {
```
I think we might want to set `conf.soft_connection_limit` to `MAX_INT` to avoid this `if`.

kazuho commented on lib/core/context.c:
```diff
@@ -187,3 +187,25 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
+    h2o_linklist_t *node, *nprev;
+    if (max_connections_to_close <= 0)
+        return 0;
+    for (node = ctx->_inactive_conns.prev; node != &ctx->_inactive_conns; node = nprev) {
```
Shouldn't we iterate the list in the normal order (i.e. by using `next`)?

I assume that the elements are ordered from the oldest to the newest, because we insert connections like by doing `h2o_linklist_insert(&conn->super.ctx->_inactive_conns, &conn->super._conns)`, which inserts the element at the end of the list. Closing the ones that have spent longer time in idle state makes sense to me.

kazuho commented on lib/core/context.c:
```diff
@@ -187,3 +187,25 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
+
+        now = h2o_gettimeofday(ctx->loop);
+        if (now.tv_sec - conn->connected_at.tv_sec < min_age)
+            continue;
```
We can do a break here, assuming that we make the change above.

That is because a connection that has lasted shorter than min_age is guaranteed to have spent less than that amount of time in the idle state.
